src/crimson/os/seastore/journal.cc

```diff
@@ -721,4 +735,108 @@ Journal::scan_segment_ret Journal::scan_segment(
+	  ).safe_then([=, &cursor, &budget_used, &handler](auto md) {
+	    logger().debug(
+	      "Journal::scan_valid_records: read complete {}",
+	      cursor.offset);
+	    if (!md) {
+	      logger().debug(
+		"Journal::scan_valid_records: found invalid header at {}",
+		cursor.offset);
+	      cursor.last_valid_header_found = true;
+	      return scan_valid_records_ertr::now();
+	    } else {
+	      logger().debug(
+		"Journal::scan_valid_records: valid record read at {}",
+		cursor.offset);
+	      cursor.last_committed = paddr_t{
+		cursor.offset.segment,
+		md->first.committed_to};
+	      cursor.pending_records.emplace_back(
+		cursor.offset,
+		md->first,
+		md->second);
+	      cursor.offset.offset +=
+		md->first.dlength + md->first.mdlength;
+	      return scan_valid_records_ertr::now();
+	    }
+	  }).safe_then([=, &cursor, &budget_used, &handler] {
+	    return crimson::do_until(
+	      [=, &budget_used, &cursor, &handler] {
+		logger().debug(
+		  "Journal::scan_valid_records: valid record read, processing queue");
+		if (cursor.pending_records.empty()) {
+		  /* This is only possible if the segment is empty.
+		   * A record's last_commited must be prior to its own
+		   * location since it itself cannot yet have been committed
+		   * at its own time of submission.  Thus, the most recently
+		   * read record must always fall after cursor.last_committed */
+		  return scan_valid_records_ertr::make_ready_future<bool>(true);
+		}
+		auto &next = cursor.pending_records.front();
+		if (next.offset > cursor.last_committed) {
+		  return scan_valid_records_ertr::make_ready_future<bool>(true);
+		}
+		budget_used +=
+		  next.header.dlength + next.header.mdlength;
+		return handler(
+		  next.offset,
+		  next.header,
+		  next.mdbuffer
+		).safe_then([&cursor] {
+		  cursor.pending_records.pop_front();
+		  return scan_valid_records_ertr::make_ready_future<bool>(false);
+		});
+	      });
+	  });
+	} else {
+	  assert(!cursor.pending_records.empty());
+	  auto &next = cursor.pending_records.front();
+	  return read_validate_data(next.offset, next.header
+	  ).safe_then([=, &budget_used, &next, &cursor, &handler](auto valid) {
+	    if (!valid) {
+	      cursor.pending_records.clear();
+	      return scan_valid_records_ertr::now();
+	    }
+	    budget_used +=
+	      next.header.dlength + next.header.mdlength;
+	    return handler(
+	      next.offset,
+	      next.header,
+	      next.mdbuffer
+	    ).safe_then([&cursor] {
+	      cursor.pending_records.pop_front();
+	      return scan_valid_records_ertr::now();
+	    });
+	  });
+	}
+      }().safe_then([=, &budget_used, &cursor] {
+	return scan_valid_records_ertr::make_ready_future<bool>(
+	  cursor.is_complete() || budget_used >= budget);
+      });
+    }).safe_then([retref=std::move(retref)]() mutable -> scan_valid_records_ret {
+      return scan_valid_records_ret(
+	scan_valid_records_ertr::ready_future_marker{},
+	std::move(*retref));
```

Comment:
nit, `*retref` would suffice.

Reply:
I'm not sure that's true.  *retref won't be an rvalue ref.

Reply:
i mean,
```c++
return scan_valid_records_ertr::make_ready_future<size_t>(*retref);
```
is good enough. because `std::move(*retref)` makes user to believe that `*retref` is a heavy-weight object or it is a move-only object, so we should move it. but actually it is but the `budget_used` and it is a `size_t`.

Reply:
Ok

Reply:
Actually, I'm going to leave it as is.  std::move here is technically correct, my intention is to permit a move construction as I do not mean to leave the previous location valid.  Doing it this way is type agnostic, if we change the return type here in the future to convey something more complex, std::move will remain correct while omitting the std::move will be incorrect.